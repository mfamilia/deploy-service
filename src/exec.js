const util = require('util')
const child = require('child_process')

module.exports = util.promisify(child.exec);
