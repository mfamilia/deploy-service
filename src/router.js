import createRouter from 'find-my-way'
import exec from './exec'


const router = createRouter({ 
  ignoreTrailingSlash: true,
  defaultRoute: (req, res) => {
    res.statusCode = 404
    res.setHeader('content-type', 'application/json')
    res.end(JSON.stringify({ data: 'not found' }))
  }
})

router.on('GET', '/health', (req, res) => {
  res.setHeader('content-type', 'application/json')
  res.end(JSON.stringify({ data: 'healthy' }))
})

const execCallback = (err, stdout, stderr) => {
  if(stdout) console.log(stdout)
  if(stderr) console.log(stderr)
}

router.on('POST', '/github', async (req, res) => {
  const payload = JSON.parse(req.body)
  const { name } = payload.repository
  const dir = process.env.PROJECTS_DIR
  const path = `${dir}/${name}`

	await exec(`git -C ${path} reset --hard`, execCallback);
	await exec(`git -C ${path} clean -df`, execCallback);
	await exec(`git -C ${path} pull -f`, execCallback);
  await exec(`${path}/update.sh`, execCallback);
  
  res.setHeader('content-type', 'application/json')
  res.end(JSON.stringify({ data: 'success' }))
});

export default router;