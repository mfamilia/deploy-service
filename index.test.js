import request from 'supertest'
import low from '0http/lib/server/low'
import cero from '0http'
import router from './src/router'
import exec from './src/exec'


jest.mock('./src/exec')

describe('Service Tests', () => {
  const baseUrl = `http://localhost:${process.env.PORT}`
  const { server } = cero({ 
    server: low(),
    router
  })

  beforeEach(() => {
    exec.mockResolvedValue()
  })

  it('should successfully register service routes', (done) => {
    server.start(~~process.env.PORT, socket => {
      if (socket) {
        done()
      }
    })
  })

  it('should GET JSON response: /health', async () => {
    const resp = await request(baseUrl).get('/health')
    const json = JSON.parse(resp.text)

    expect(resp.status).toEqual(200)
    expect(json.data).toEqual('healthy')
  })

  it('should GET 404 response: /missing', async () => {
    const resp = await request(baseUrl).get('/missing')
    const json = JSON.parse(resp.text)

    expect(resp.status).toEqual(404)
    expect(json.data).toEqual('not found')
  })

  it('should GET 404 response: /bad/%url', async () => {
    const resp = await request(baseUrl).get('/bad/%url')
    const json = JSON.parse(resp.text)

    expect(resp.status).toEqual(404)
    expect(json.data).toEqual('not found')
  })

  it('should POST JSON: /github', async () => {
    const resp = await request(baseUrl)
      .post('/github')
      .send({ repository: { name: 'sample' }})

    const json = JSON.parse(resp.text)

    expect(resp.status).toEqual(200)
    expect(json.data).toEqual('success')
    expect(exec.mock.calls).toMatchSnapshot()
  })

  it('should successfully terminate the service', async () => {
    server.close()
  })
})