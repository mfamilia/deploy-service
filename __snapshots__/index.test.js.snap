// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Service Tests should POST JSON: /github 1`] = `
Array [
  Array [
    "git -C ./projects/sample reset --hard",
    [Function],
  ],
  Array [
    "git -C ./projects/sample clean -df",
    [Function],
  ],
  Array [
    "git -C ./projects/sample pull -f",
    [Function],
  ],
  Array [
    "./projects/sample/update.sh",
    [Function],
  ],
]
`;
